src/sage/rings/polynomial/multi_polynomial_sequence.py

```diff
@@ -943,13 +943,24 @@ def connected_components(self):
              Polynomial Sequence with 128 Polynomials in 128 Variables,
              Polynomial Sequence with 128 Polynomials in 128 Variables]
         """
-        g = self.connection_graph()
-        C = sorted(g.connected_components())
-
-        P = [[] for _ in range(len(C))]
+        # Use a union-find data structure to encode relationships between
+        # variables, i.e., that they belong to a same polynomial
+        from sage.sets.disjoint_set import DisjointSet
+        DS = DisjointSet(self.variables())
```

Comment:
```suggestion
        vs = [f.variables() for f in self]
        DS = DisjointSet(set().union(*vs))
```

Reply:
Rather than `self.variables()` I think than one should use `self.universe().variables()`. The point being that the vertices should be all variables in the underlying polynomial ring (and not just the ones that appear in some polynomial in `self`). Doing this change will also be better in terms of efficiency.

Reply:
Actually, `self.universe().variables()` is only available for boolean polynomial rings. One should use `self.universe().gens()` for other polynomial ring (and possibly fix this discrepency between boolean and other polynomial rings).